### PR TITLE
[Dev] Enable dense routing maps for Flex dispatch

### PR DESCRIPTION
**Pull request by harryzhou2000**
## Rebase update (2026-09-07)

Rebased the authentic `dev` PR onto current `dev` at `0f9c777f7` (`[dev] [DeepSeek-V4] Add compact BF16 and MXFP8 DSA indexer support (#5992)`). The eight-commit series rebased cleanly apart from one test-file conflict caused by new adjacent checkpoint tests; the resolution preserves both current `dev` coverage and this PR's runtime/checkpoint HybridEP routing-mode cases. New head: `2ea7c0487`.

Post-rebase audit:

- No unresolved GitHub review threads.
- `git diff --check`, changed-file Ruff, and Python byte-compilation pass.
- The final diff remains scoped to dense routing and its tests; no dependency or lockfile changes.
- `moe_hybridep_routing_map_mode: "indices"` remains in `GOLDEN_CONFIG`.

Local validation used `nvcr.io/nvidia/pytorch:26.04-py3` on 2x NVIDIA A100 80GB PCIe with the companion rebased TE source/native extension:

- Checkpoint routing-mode cases: 3 passed on each of 2 ranks.
- Two-GPU-compatible focused router cases: 10 passed on each rank.
- Focused dispatcher metadata/API cases: 9 passed on each rank.
- Direct A100 integration equivalence passed for TE int16 and int64 dense output, with and without Quantile Balancing, including routing probabilities, route identities, backward gradients, and QB histograms.
- Dependency audit: no broken requirements.

The real transport and expert-bias additions whose existing fixtures explicitly initialize EP=8 could not be rerun on this 2-GPU host; attempting the latter reached the expected topology guard (`world_size (2) is not divisible by ... (8)`) before the test body. They retain the earlier EP8 validation below and remain intended for functional CI.

- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

Enable MCore Flex dispatchers to consume Transformer Engine's direct dense top-k routing indices whenever the selected backend requires dense metadata and the installed TE exposes the optional `topk_indices` output buffer.

DeepEP v1, DeepEP v2, and NCCL-EP now request `int64 [num_tokens, topk]` indices automatically. HybridEP now uses `moe_hybridep_routing_map_mode`, which defaults to `"indices"`; set it to `"bool"` to force the boolean map. Index mode uses `int16` only when the installed backend reports dense-routing support and falls back to bool above the `int16` expert limit or when support is unavailable. Non-Flex dispatchers and older TE combinations preserve the bool router-output path, and dispatcher compatibility paths reconstruct indices when a dense consumer requires them.

The dispatcher managers accept both representations, so direct TE output avoids reconstructing indices with `torch.topk` while compatibility paths remain intact. Dense padding routes remain `-1`, expert-TP expansion preserves invalid routes, and expert-bias counts are accumulated from the selected indices.

Companion main PR: #6615.

## Previous testing

- Checkpoint argument restoration: 2 passed (`"bool"` is restored; a missing legacy field keeps `"indices"`).
- Focused router, routing-map, dispatcher-metadata, padding, and feature-detection coverage: 27 passed.
- Real HybridEP EP=8 forward/backward passed for both `moe_hybridep_routing_map_mode="bool"` and `"indices"` using upstream HybridEP. The indices case validates the direct dense `int16 [num_tokens, topk]` path from TE through MCore into HybridEP.
- Dense expert-bias accounting passed in deterministic and non-deterministic modes, including fused dense routing.
- `git diff --check`, Ruff, and `py_compile` passed.

Real DeepEP v1, DeepEP v2, and NCCL-EP transport were not rerun in this validation. Their dense representation selection and compatibility-manager paths remain covered by focused tests.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: N/A

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.nvidia.com/megatron-core/developer-guide/latest/developer-guide/advanced.html#typing)
- [x] I have added relevant documentation
- [ ] I have run the autoformatter on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

When the PR is ready, mark it ready for review after conflicts are resolved and CI is passing.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied automatically and final reviewers are assigned.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied automatically.

### Merge

Any member of `mcore-engineers` will be able to merge the PR.

### CODE REVIEW

File: megatron/core/transformer/moe/fused_a2a.py [hunk 1]
```diff
@@ -636,7 +662,25 @@ def forward(
         # If we provide the num_permuted_tokens, we do not need to use sync to
         # wait for the data in pinned memory ready
         non_blocking = num_permuted_tokens is not None
-        # Process the dispatch
+        if topk_idx is not None and not HAVE_HYBRIDEP_DENSE_ROUTING:
```
Comment (Autumn1998) [inline megatron/core/transformer/moe/fused_a2a.py]: Maybe we can remove the check `if topk_idx is not None and not HAVE_HYBRIDEP_DENSE_ROUTING:` and add `"assert routing_map is not None"` on the `else` branch of `if use_dense: xxx else`

Comment (harryzhou2000) [reply megatron/core/transformer/moe/fused_a2a.py]: Good point, thanks. I removed the redundant unsupported-dense check and now assert that `routing_map` is available in the sparse branch. This lets the adapter fall back to sparse routing when dense metadata is unsupported but a sparse map is available, while still failing immediately if neither representation can be used. Applied in 737cf1de0 and mirrored to the companion main PR #6615 in bd5e540a1.

File: megatron/core/transformer/moe/router.py [hunk 2]
```diff
@@ -718,13 +749,27 @@ def _apply_expert_bias(
...
+                    if use_dense_indices:
+                        routing_map = routing_map[~flat_mask]
+                    else:
+                        routing_map = routing_map & (~flat_mask).unsqueeze(-1)
+                if use_dense_indices:
+                    expert_indices = routing_map.reshape(-1).to(torch.long)
+                    token_counts = torch.ones_like(
+                        expert_indices, dtype=self.local_tokens_per_expert.dtype
+                    )
+                    if torch.are_deterministic_algorithms_enabled():
```
Comment (YangFei1990) [inline megatron/core/transformer/moe/router.py]: Could we have some comment for this logic?

**Review by FDecaYed — APPROVED** [review]
Reviewed the current dense-routing implementation and compatibility paths. The feature gating, sparse fallbacks, expert-TP expansion, checkpoint behavior, and focused coverage look consistent.

**Comment by chtruong814** [comment]
@FDecaYed @harryzhou2000 cc: @balasaajay This PR has conflicts that will need to be resolved.  I cancelled the CI run for now otherwise, it would need to be re-run again anyway.  Please resolve and merge main into this branch.

**Comment by FDecaYed** [comment]
This PR now conflicts with its `dev` base. The earlier CI was canceled during the merge-base/queue transition; it was not a code failure. Could you rebase `hhanyu/dense-routing-dev` onto the current `dev` rather than `main`, while preserving the `moe_hybridep_routing_map_mode` `GOLDEN_CONFIG` fix? Once the new SHA is up, I can re-review it and restart functional CI.
